src/js/diff.js

```diff
@@ -0,0 +1,549 @@
+            removedSegment[key] = leftValue;
+            segments.push(removedSegment);
+        }
+        if ((!isArray && rightValue !== undefined) || (isArray && rightValue.length)) {
+            var addedSegment = { type: "added"};
+            addedSegment[key] = rightValue;
+            segments.push(addedSegment);
+        }
+    }
+    return segments;
+};
+
+/**
+ *
+ * Find the longest matching "phrase" in two strings, which prefers one or more whole words.
+ *
+ * @param leftString {String} - One of the two strings, which will be used to search for matching patterns.
+ * @param rightString {String} - The other string, which will be compared against all possible sub-patterns.
+ * @return {Object} - See below.
+ *
+ * The object returned has three keys:
+ *
+ * `leftIndex`:  The position of the matching segment in `leftString`.  If there is no matching segment, this will be `-1`.
+ * `rightIndex`: The position of the matching segment in `rightString`.  If there is no matching segment, this will be `-1`.
+ * `segment`:    The longest matching string common to both strings.  If there are multiple matching strings of the same length, the first is returned.
+ *
+ */
+gpii.diff.longestCommonStringPhrase = function (leftString, rightString) {
+    var matchDef = {
+        leftIndex: -1,
+        rightIndex: -1,
+        segment: ""
+    };
+
+    if (leftString && rightString) {
+        var leftWordPhraseDefs = gpii.diff.extractPhraseSegments(leftString);
+
+        var firstMatchDef = fluid.find(leftWordPhraseDefs, function (phraseDef) {
+            var rightIndex = rightString.indexOf(phraseDef.value);
+            if (rightIndex !== -1) {
+                return {
+                    segment: phraseDef.value,
+                    leftIndex: phraseDef.index,
+                    rightIndex: rightIndex
+                };
+            }
+        });
+        if (firstMatchDef) {
+            matchDef = firstMatchDef;
+        }
+    }
+    return matchDef;
+};
+
+gpii.diff.sortPhraseDefsByLength = function (a, b) {
+    if (a.value.length > b.value.length) {
+        return -1;
+    }
+    else if (a.value.length === b.value.length) {
+        if (a.index < b.index) {
+            return -1;
+        }
+        else if (a.index === b.index) {
+            return 0;
+        }
+        else if (b.index < a.index) {
+            return 1;
+        }
+    }
+    else {
+        return 1;
+    }
+};
+
+/**
+ *
+ * Break down a string into an array of "phrase segments", ordered from longest (the original string) to shortest (the
+ * shortest individual word).  Each "phrase segment" looks like:
+ * {
+ *  value: "a word or phrase, possibly including whitespace ",
+ *  index: 0 // The position of the phrase in `originalString`
+ * }
+ *
+ * @param originalString {String} - The original string to break down.
+ * @returns {Array} - An array of "phrase segments".
+ *
+ */
+gpii.diff.extractPhraseSegments = function (originalString) {
```

Comment:
There's enough code in your "extractPhraseSegments", "compareStrings", "compareArrays" and "longestCommonArraySegment" functions combined to at least consider the option of replacing them all with the standard LCS algorithm - on the plus side this would make this package suitable for differencing more substantial quantities of data and perhaps rationalise the implementation somewhat - on the minus side are the cognitive costs of maintaining, comprehending the algorithm and its various corner cases. I found an impl which has at least one test case (!) here: https://github.com/erictrinh/lcs/blob/master/index.js
I think if we don't, we should at least advertise that we are not using this standard algorithm in the various corresponding places in the README since I think it has become an expectation for someone using a utility named "diff" which operates on sequence-based data.

Reply:
The approach described in the article and implemented in the linked repo unfortunately looks for the longest common subsequence irrespective of whether the members are adjacent.  So, when comparing the strings "abc" and "acd", that library returns "ac" as the longest common sub-segment.

One option would be to use an approach from [the related article on substrings](https://en.wikipedia.org/wiki/Longest_common_substring_problem, which I'll need to review for a  bit.

We might also (as I've done in previous approaches) break a string down into segments of blocks of word characters interleaved with blocks of non-word characters.  Once we have that, we can use the "standard" array comparison.

To use this, we'd need a little more than just the longest common sequence.  We need to know the position of each member of the sequence in each compared sequence, so that we can represent the removed/added material in context.

We also need to know how ties are broken, as there may be multiple differing longest sequences, as when comparing "xaybzc" and "abcxyz".  It's not 100% clear to me if the "standard" approach is "longest first" or not, or uses another mechanism to break ties.

Reply:
Yes, this is a important subtlety to note - it looks like we do indeed need two algorithms rather than one. The good news is that the substring problem is substantially easier than the subsequence problem and can be done in linear time.